File: packages/api/src/domain/household.ts
import { asDateTimeString, asNonEmptyString, asString, isRecord } from '../lib/normalize'

export type HouseholdRole = 'owner' | 'member'

export interface HouseholdMember {
  userId: string
  email: string
  /** Display name; null when never set (consumers fall back to email). */
  displayName: string | null
  role: HouseholdRole
  joinedAt: string
}

export interface Household {
  id: string
  createdAt: string
  members: HouseholdMember[]
}

const isHouseholdRole = (value: unknown): value is HouseholdRole =>
  value === 'owner' || value === 'member'

export const normalizeHouseholdMember = (value: unknown): HouseholdMember | null => {
  if (!isRecord(value)) {
    return null
  }

  const userId = asNonEmptyString(value.userId)
  const email = asNonEmptyString(value.email)
  const displayName = value.displayName === null ? null : asString(value.displayName)
  const role = isHouseholdRole(value.role) ? value.role : null
  const joinedAt = asDateTimeString(value.joinedAt)

  if (!userId || !email || displayName === null || !role || !joinedAt) {
    return null
  }

  return { userId, email, displayName, role, joinedAt }
}

export const normalizeHousehold = (value: unknown): Household | null => {
  if (!isRecord(value) || !Array.isArray(value.members)) {
    return null
  }

  const id = asNonEmptyString(value.id)
  const createdAt = asDateTimeString(value.createdAt)
  if (!id || !createdAt) {
    return null
  }

  const members: HouseholdMember[] = []
  for (const raw of value.members) {
    const member = normalizeHouseholdMember(raw)
    if (!member) {
      return null
    }
    members.push(member)
  }

  return { id, createdAt, members }
}

File: backend/internal/domain/household.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// HouseholdRole is a member's role in a household. Roles are stored in v1 but
// carry no behavioral split yet (ADR-0002); both roles see and modify all of
// the household's data equally.
type HouseholdRole string

const (
	HouseholdRoleOwner  HouseholdRole = "owner"
	HouseholdRoleMember HouseholdRole = "member"
)

// Household is the shared data space: every account, category, transaction,
// debtor, debt operation, and planned payment belongs to exactly one
// household, and members access it equally.
type Household struct {
	ID        uuid.UUID
	CreatedAt time.Time
	Members   []HouseholdMember
}

// HouseholdMember is one user's membership in a household, joined with the
// profile fields the member listing exposes. DisplayName is nil when never
// set (consumers fall back to Email).
type HouseholdMember struct {
	UserID      uuid.UUID
	Email       string
	DisplayName *string
	Role        HouseholdRole
	JoinedAt    time.Time
}

// Membership is a user's single (v1) household membership row - the auth
// middleware's household resolution input.
type Membership struct {
	HouseholdID uuid.UUID
	UserID      uuid.UUID
	Role        HouseholdRole
	JoinedAt    time.Time
}
